File: avisena-backend-copy/app/crud/sensors.py
from sqlalchemy.orm import Session
from sqlalchemy import text
from sqlalchemy.exc import SQLAlchemyError
from typing import Optional
import logging
from app.schemas.sensors import SensorCreate, SensorUpdate

logger = logging.getLogger(__name__)

def create_sensor(db: Session, sensor: SensorCreate) -> Optional[bool]:
    try:
        sentencia = text("""
            INSERT INTO sensores(
                nombre, id_tipo_sensor, id_galpon, descripcion, estado
            ) VALUES (
                :nombre, :id_tipo_sensor, :id_galpon, :descripcion, :estado
            )
        """)
        db.execute(sentencia, sensor.model_dump())
        db.commit()
        return True
    except SQLAlchemyError as e:
        db.rollback()
        logger.error(f"Error al crear sensor: {e}")
        raise Exception("Error de base de datos al crear el sensor")


def get_sensor_by_id(db: Session, id_sensor: int):
    try:
        query = text("""
            SELECT 
                s.id_sensor, s.nombre, s.id_tipo_sensor, s.id_galpon, s.descripcion, s.estado,
                t.nombre AS nombre_tipo, t.modelo AS modelo_tipo, g.nombre AS nombre_galpon
            FROM sensores s
            INNER JOIN tipo_sensores t ON s.id_tipo_sensor = t.id_tipo
            INNER JOIN galpones g ON s.id_galpon = g.id_galpon
            WHERE s.id_sensor = :id
        """)
        result = db.execute(query, {"id": id_sensor}).mappings().first()
        return result
    except SQLAlchemyError as e:
        logger.error(f"Error al obtener sensor: {e}")
        raise Exception("Error de base de datos al obtener el sensor")


def get_all_sensores(db: Session):
    try:
        query = text("""
            SELECT 
                s.id_sensor, s.nombre, s.id_tipo_sensor, s.id_galpon, s.descripcion, s.estado,
                t.nombre AS nombre_tipo, t.modelo AS modelo_tipo, g.nombre AS nombre_galpon
            FROM sensores s
            INNER JOIN tipo_sensores t ON s.id_tipo_sensor = t.id_tipo
            INNER JOIN galpones g ON s.id_galpon = g.id_galpon
            ORDER BY s.nombre
        """)
        result = db.execute(query).mappings().all()
        return result
    except SQLAlchemyError as e:
        logger.error(f"Error al obtener sensores: {e}")
        raise Exception("Error de base de datos al obtener los sensores")


def get_sensores_by_galpon(db: Session, id_galpon: int):
    try:
        query = text("""
            SELECT 
                s.id_sensor, s.nombre, s.id_tipo_sensor, s.id_galpon, s.descripcion, s.estado,
                t.nombre AS nombre_tipo, t.modelo AS modelo_tipo, g.nombre AS nombre_galpon
            FROM sensores s
            INNER JOIN tipo_sensores t ON s.id_tipo_sensor = t.id_tipo
            INNER JOIN galpones g ON s.id_galpon = g.id_galpon
            WHERE s.id_galpon = :id_galpon
            ORDER BY s.nombre
        """)
        result = db.execute(query, {"id_galpon": id_galpon}).mappings().all()
        return result
    except SQLAlchemyError as e:
        logger.error(f"Error al obtener sensores por galpón: {e}")
        raise Exception("Error de base de datos al obtener los sensores")


def update_sensor_by_id(db: Session, id_sensor: int, sensor: SensorUpdate) -> Optional[bool]:
    try:
        sensor_data = sensor.model_dump(exclude_unset=True)
        if not sensor_data:
            return False

        set_clauses = ", ".join([f"{key} = :{key}" for key in sensor_data.keys()])
        sentencia = text(f"""
            UPDATE sensores 
            SET {set_clauses}
            WHERE id_sensor = :id_sensor
        """)

        sensor_data["id_sensor"] = id_sensor
        result = db.execute(sentencia, sensor_data)
        db.commit()

        return result.rowcount > 0
    except SQLAlchemyError as e:
        db.rollback()
        logger.error(f"Error al actualizar sensor {id_sensor}: {e}")
        raise Exception("Error de base de datos al actualizar el sensor")


def change_sensor_status(db: Session, id_sensor: int, nuevo_estado: bool) -> bool:
    try:
        sentencia = text("""
            UPDATE sensores
            SET estado = :estado
            WHERE id_sensor = :id_sensor
        """)
        result = db.execute(sentencia, {"estado": nuevo_estado, "id_sensor": id_sensor})
        db.commit()
        return result.rowcount > 0
    except SQLAlchemyError as e:
        db.rollback()
        mensaje_error = str(e)
        if "45000" in mensaje_error or "No se puede activar" in mensaje_error:
            raise Exception("No se puede activar un sensor de un tipo de sensor inactivo")
        raise Exception("Error de base de datos al cambiar el estado del sensor")

File: assets/js/sensors.js
import { sensorService } from "../api/sensor.service.js";
import { tipoSensorService } from "../api/sensorType.service.js";
import { shedService } from "../api/shed.service.js";

let modalInstance = null;
let createModalInstance = null;
let allSensors = [];
let allTipos = [];
let allGalpones = [];

function createSensorRow(sensor) {
  const sensorId = sensor.id_sensor;

  return `
    <tr>
      <td class="cell">${sensor.nombre}</td>
      <td class="cell">${sensor.nombre_tipo} - ${sensor.modelo_tipo}</td>
      <td class="cell">${sensor.nombre_galpon}</td>
      <td class="cell">${sensor.descripcion}</td>
      <td class="cell">
        <div class="form-check form-switch d-inline-block">
          <input class="form-check-input sensor-status-switch" type="checkbox" role="switch" 
                 id="switch-sensor-${sensorId}" data-sensor-id="${sensorId}" 
                 ${sensor.estado ? 'checked' : ''}>
          <label class="form-check-label" for="switch-sensor-${sensorId}">
            ${sensor.estado ? 'Activo' : 'Inactivo'}
          </label>
        </div>
      </td>
      <td class="cell">
        <button class="btn btn-sm btn-success btn-edit-sensor" data-sensor-id="${sensorId}">
          <i class="fa-regular fa-pen-to-square"></i>
        </button>
      </td>
    </tr>
  `;
}

function filterSensors() {
  const estadoFilter = document.getElementById("filter-sensor-estado").value;
  const tipoFilter = document.getElementById("filter-tipo").value;
  const galponFilter = document.getElementById("filter-galpon").value;

  let filteredSensors = allSensors;

  if (estadoFilter === "active") {
    filteredSensors = filteredSensors.filter(s => s.estado === 1 || s.estado === true);
  } else if (estadoFilter === "inactive") {
    filteredSensors = filteredSensors.filter(s => s.estado === 0 || s.estado === false);
  }

  if (tipoFilter !== "all") {
    filteredSensors = filteredSensors.filter(s => s.id_tipo_sensor == tipoFilter);
  }

  if (galponFilter !== "all") {
    filteredSensors = filteredSensors.filter(s => s.id_galpon == galponFilter);
  }

  const tableBody = document.getElementById("sensors-table-body");
  if (!tableBody) return;

  if (filteredSensors.length > 0) {
    tableBody.innerHTML = filteredSensors.map(createSensorRow).join("");
  } else {
    tableBody.innerHTML = `
      <tr>
        <td colspan="6" class="text-center">No hay sensores con esos filtros.</td>
      </tr>
    `;
  }
}

async function cargarFiltros() {
  try {
    allTipos = await tipoSensorService.getTipoSensores();
    const filterTipo = document.getElementById("filter-tipo");
    if (filterTipo) {
      filterTipo.innerHTML = '<option value="all">Todos los tipos</option>';
      allTipos.forEach((t) => {
        filterTipo.innerHTML += `<option value="${t.id_tipo}">${t.nombre} - ${t.modelo}</option>`;
      });
    }

    allGalpones = await shedService.getGalponesActivos();
    const filterGalpon = document.getElementById("filter-galpon");
    if (filterGalpon) {
      filterGalpon.innerHTML = '<option value="all">Todos los galpones</option>';
      allGalpones.forEach((g) => {
        filterGalpon.innerHTML += `<option value="${g.id_galpon}">${g.nombre}</option>`;
      });
    }
  } catch (error) {
    console.error("Error al cargar filtros:", error);
  }
}

async function cargarTiposSensores() {
  try {
    const tipos = await tipoSensorService.getTipoSensoresActivos();

    const selectCrear = document.getElementById("create-sensor-id_tipo_sensor");
    if (selectCrear) {
      selectCrear.innerHTML =
        '<option value="" disabled selected>Seleccione un tipo</option>';

      tipos.forEach((t) => {
        if (t.id_tipo !== null && t.id_tipo !== undefined)
          selectCrear.innerHTML += `<option value="${t.id_tipo}">${t.nombre} - ${t.modelo}</option>`;
      });
    }

    const selectEditar = document.getElementById("edit-sensor-id_tipo_sensor");
    if (selectEditar) {
      selectEditar.innerHTML =
        '<option value="" disabled selected>Seleccione un tipo</option>';

      tipos.forEach((t) => {
        if (t.id_tipo !== null && t.id_tipo !== undefined)
          selectEditar.innerHTML += `<option value="${t.id_tipo}">${t.nombre} - ${t.modelo}</option>`;
      });
    }
  } catch (error) {
    console.error("Error al cargar tipos de sensores:", error);
  }
}

async function cargarGalpones() {
  try {
    const galpones = await shedService.getGalponesActivos();

    const selectCrear = document.getElementById("create-sensor-id_galpon");
    if (selectCrear) {
      selectCrear.innerHTML =
        '<option value="" disabled selected>Seleccione un galpón</option>';
      galpones.forEach((g) => {
        selectCrear.innerHTML += `<option value="${g.id_galpon}">${g.nombre}</option>`;
      });
    }

    const selectEditar = document.getElementById("edit-sensor-id_galpon");
    if (selectEditar) {
      selectEditar.innerHTML =
        '<option value="" disabled selected>Seleccione un galpón</option>';
      galpones.forEach((g) => {
        selectEditar.innerHTML += `<option value="${g.id_galpon}">${g.nombre}</option>`;
      });
    }
  } catch (error) {
    console.error("Error al cargar galpones:", error);
  }
}

async function openEditModal(sensorId) {
  const modalElement = document.getElementById("edit-sensor-modal");
  if (!modalInstance){
    modalInstance = new bootstrap.Modal(modalElement);
  }
  try {
    const sensor = await sensorService.getSensorById(sensorId);
    document.getElementById("edit-sensor-id").value = sensor.id_sensor;
    document.getElementById("edit-sensor-nombre").value = sensor.nombre;
    document.getElementById("edit-sensor-id_tipo_sensor").value = sensor.id_tipo_sensor;
    document.getElementById("edit-sensor-id_galpon").value = sensor.id_galpon;
    document.getElementById("edit-sensor-descripcion").value = sensor.descripcion;
    modalInstance.show();
  } catch (error) {
    console.error("Error al obtener sensor:", error);
    Swal.fire({
      icon: "error",
      title: "Error",
      text: "No se pudieron cargar los datos del sensor.",
    });
  }
}

async function handleUpdateSubmit(event) {
  event.preventDefault();

  const sensorId = document.getElementById("edit-sensor-id").value;
  const descripcion = document.getElementById("edit-sensor-descripcion").value;

  if (descripcion.length < 10) {
    Swal.fire({
      icon: "warning",
      title: "Descripción Inválida",
      text: "La descripción debe tener al menos 10 caracteres.",
    });
    return;
  }

  const updatedData = {
    nombre: document.getElementById("edit-sensor-nombre").value,
    id_tipo_sensor: Number(document.getElementById("edit-sensor-id_tipo_sensor").value),
    id_galpon: Number(document.getElementById("edit-sensor-id_galpon").value),
    descripcion: document.getElementById("edit-sensor-descripcion").value,
  };

  try {
    await sensorService.updateSensor(sensorId, updatedData);
    modalInstance.hide();

    Swal.fire({
      icon: "success",
      title: "Éxito",
      text: "Sensor actualizado exitosamente.",
    });

    init();
  } catch (error) {
    console.error("Error al actualizar sensor:", error);

    Swal.fire({
      icon: "error",
      title: "Actualizar",
      text: "No se pudo actualizar el sensor.",
    });
  }
}

async function handleCreateSubmit(event) {
  event.preventDefault();

  const nombre = document.getElementById("create-sensor-nombre").value.trim();
  const id_tipo_sensor = Number(document.getElementById("create-sensor-id_tipo_sensor").value);
  const id_galpon = Number(document.getElementById("create-sensor-id_galpon").value);
  const descripcion = document.getElementById("create-sensor-descripcion").value.trim();

  if (descripcion.length < 10) {
    Swal.fire({
      icon: "warning",
      title: "Descripción Inválida",
      text: "La descripción debe tener al menos 10 caracteres.",
    });
    return;
  }

  const newSensorData = {
    nombre,
    id_tipo_sensor,
    id_galpon,
    descripcion,
    estado: true,
  };

  try {
    await sensorService.createSensor(newSensorData);

    if (createModalInstance) createModalInstance.hide();
    document.getElementById("create-sensor-form").reset();

    Swal.fire({
      icon: "success",
      title: "Éxito",
      text: "Sensor creado exitosamente.",
    });

    init();
  } catch (error) {
    console.error("Error al crear sensor:", error);

    Swal.fire({
      icon: "error",
      title: "Error",
      text: "No se pudo crear el sensor.",
    });
  }
}

async function handleStatusSwitch(event) {
  const switchElement = event.target;
  if (!switchElement.classList.contains("sensor-status-switch")) return;

  const sensorId = switchElement.dataset.sensorId;
  const newStatus = switchElement.checked;

  Swal.fire({
    title: "¿Está seguro?",
    text: `¿Desea ${newStatus ? "activar" : "desactivar"} este sensor?`,
    icon: "question",
    showCancelButton: true,
    confirmButtonText: "Sí, confirmar",
    cancelButtonText: "Cancelar",
  }).then(async (result) => {
    if (result.isConfirmed) {
      try {
        await sensorService.changeSensorStatus(sensorId, newStatus);

        Swal.fire({
          icon: "success",
          title: "Éxito",
          text: `El sensor ha sido ${newStatus ? "activado" : "desactivado"}.`,
        });

        init();
      } catch (error) {
        console.error("Error al cambiar estado:", error);
        Swal.fire({
          icon: "error",
          title: "Error",
          text: "No se pudo cambiar el estado.",
        });
        switchElement.checked = !newStatus;
      }
    } else {
      switchElement.checked = !newStatus;
    }
  });
}

async function handleTableClick(event) {
  const editButton = event.target.closest(".btn-edit-sensor");
  if (editButton) {
    const sensorId = editButton.dataset.sensorId;
    openEditModal(sensorId);
    return;
  }
}

function handleFilterChange() {
  filterSensors();
}

async function init() {
  const tableBody = document.getElementById("sensors-table-body");
  if (!tableBody) return;

  const colspan = 6;
  tableBody.innerHTML =
    `<tr><td colspan="${colspan}" class="text-center">Cargando sensores</td></tr>`;

  const createModalElement = document.getElementById("create-sensor-modal");
  if (createModalElement && !createModalInstance) {
    createModalInstance = new bootstrap.Modal(createModalElement);
  }

  await cargarFiltros();
  await cargarTiposSensores();
  await cargarGalpones();

  try {
    const sensors = await sensorService.getSensors();
    allSensors = sensors || [];

    if (allSensors.length > 0) {
      tableBody.innerHTML = allSensors.map(createSensorRow).join("");
    } else {
      tableBody.innerHTML =
        `<tr><td colspan="${colspan}" class="text-center">No hay sensores registrados.</td></tr>`;
    }
  } catch (error) {
    console.error("Error al cargar sensores:", error);
    tableBody.innerHTML =
      `<tr><td colspan="${colspan}" class="text-center text-danger">Error al cargar datos.</td></tr>`;
  }

  const editForm = document.getElementById("edit-sensor-form");
  const createForm = document.getElementById("create-sensor-form");
  const filterEstado = document.getElementById("filter-sensor-estado");
  const filterTipo = document.getElementById("filter-tipo");
  const filterGalpon = document.getElementById("filter-galpon");

  tableBody.removeEventListener("click", handleTableClick);
  tableBody.addEventListener("click", handleTableClick);

  tableBody.removeEventListener("change", handleStatusSwitch);
  tableBody.addEventListener("change", handleStatusSwitch);

  editForm.removeEventListener("submit", handleUpdateSubmit);
  editForm.addEventListener("submit", handleUpdateSubmit);

  createForm.removeEventListener("submit", handleCreateSubmit);
  createForm.addEventListener("submit", handleCreateSubmit);

  if (filterEstado) {
    filterEstado.removeEventListener("change", handleFilterChange);
    filterEstado.addEventListener("change", handleFilterChange);
  }

  if (filterTipo) {
    filterTipo.removeEventListener("change", handleFilterChange);
    filterTipo.addEventListener("change", handleFilterChange);
  }

  if (filterGalpon) {
    filterGalpon.removeEventListener("change", handleFilterChange);
    filterGalpon.addEventListener("change", handleFilterChange);
  }
}

document.addEventListener("sensorTypesUpdated", async (e) => {
  try {
    await cargarFiltros();
    await cargarTiposSensores();
    const sensors = await sensorService.getSensors();
    allSensors = sensors || [];
    filterSensors();
  } catch (err) {
    console.error("Error al procesar sensorTypesUpdated:", err);
  }
});

export { init };
